### PR TITLE
Fix/meta 3037 pending tasks

### PR DESCRIPTION
**Pull request by BurakAltas**
## Change description

> In case of pending tasks, classifications could not be deleted. So, we delete the pending task since the propagation is not completed.

## Type of change
- [x] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### Code review 

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [x] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 1]
```diff
@@ -2809,12 +2810,21 @@ public void deleteClassification(String entityGuid, String classificationName) t
                             .anyMatch(x -> classificationHasPendingTask(x, classificationVertexId, entityGuid));
 
                     if (pendingTaskExists) {
-                        LOG.error("Another tag propagation is in queue for classification: {} and entity: {}. Please try again", classificationVertexId, entityGuid);
-                        throw new AtlasBaseException(AtlasErrorCode.DELETE_TAG_PROPAGATION_NOT_ALLOWED, classificationVertexId, entityGuid);
+                        List<AtlasTask> entityClassificationPendingTasks = entityPendingTasks.stream()
+                                .filter(t -> t.getParameters().get("entityGuid").equals(entityGuid)
+                                        && t.getParameters().get("classificationVertexId").equals(classificationVertexId))
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2815]: Certain types of tasks might not have entityGuid, classificationVertexId defined in parameters, which will result in an NPE.
Add check for task type as well - CLASSIFICATION_PROPAGATION_ADD

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 2]
```diff
@@ -2809,12 +2810,21 @@ public void deleteClassification(String entityGuid, String classificationName) t
...
+                                .collect(Collectors.toList());
+                        for (AtlasTask entityClassificationPendingTask: entityClassificationPendingTasks) {
+                            String taskGuid = entityClassificationPendingTask.getGuid();
+                            taskManagement.deleteByGuid(taskGuid);
+                            AtlasGraphUtilsV2.deleteProperty(entityVertex, PENDING_TASKS_PROPERTY_KEY, taskGuid);
+                            propagateDelete = false;
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2821]: There are some other places from unnecessary ADD tasks might get created, until we fix that let's keep propagateDelete as true.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java (hunk 1 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2815]: When I add check for task type CLASSIFICATION_PROPAGATION_ADD, NPE situation is fixed on its own or do I handle it separately? I mean, do all CLASSIFICATION_PROPAGATION_ADD types have entityGuid and classificationVertexId?

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java [hunk 3]
```diff
@@ -204,6 +204,32 @@ public void complete(AtlasVertex taskVertex, AtlasTask task) {
...
+                setEncodedProperty(taskVertex, Constants.TASK_UPDATED_TIME, System.currentTimeMillis());
+            }
+        }
+        catch (Exception exception) {
+            LOG.error("Error: on soft delete: {}", guid);
+
+            throw new AtlasBaseException(exception);
+        }
+        finally {
+            graph.commit();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:229]: I don't think we need manually commit the changes.
Can you please verify this once
Also check GraphTransactionInterceptor class

File: repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java [hunk 4]
```diff
@@ -205,6 +205,14 @@ public void deleteByGuids(List<String> guids) throws AtlasBaseException {
         }
     }
 
+    public void softDelete(String guid) throws AtlasBaseException{
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java:208]: Change method name to deleteByGuid and add another param -> deleteType which can be HARD/SOFT.

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java (hunk 3 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java]: You are right, doCommit automatically invoked

**Pull request closed without merging**